### ISSUE TITLE
msg_name_to_snake_case and service names

### ISSUE THREAD

**Issue by tsloughter** [tomas-abrahamsson/gpb#122]
I'm still toying around with how to best do the mapping with grpc but I do think this is at least a confusing name for the option, whether or not the user will actually want the server and method names to be snake case.

What do you think of a the `msg_name` option only changing messages (and enums) and not service/methods and having `all_to_snake_case` as a new option to change everything?

Relatedly, `use_packages` with `msg_name_to_snake_case` results in `'grpc.testing._streaming_output_call_request'`. It should likely drop the trailing `_` if there isn't a prefix. Could also consider `all_to_snake_case` resulting in snaking the package name to give `grpc_testing_streaming_output_call_request`.

**Comment by tomas-abrahamsson** [comment]
Sorry about the delay, will look into it. Ack on the leading underscore, though.

**Comment by tomas-abrahamsson** [comment]
What is exposed over grpc? Both service name and rpc name? I'm thinking also that the same needs, wrt services/rpcs and grpc, may exist for all name transforms (of which probably only snake casing is the only really Erlangy and therefore the most relevant one)

**Comment by tsloughter** [comment]
Package, service and rpc name. 

```
Service-Name → ?( {proto package name} “.” ) {service name}
Message-Type → {fully qualified proto message name}
```

So a proto like:

```protobuf
package routeguide;

service RouteGuide {
  rpc GetFeature(...) returns ... {};
}
```

Then the http2 POST path will be `routeguide.RouteGuide/GetFeature`.

So what I'm doing is my grpc library and template for generating behaviours for callbacks is still snake casing the service and methods but not with gpb.

I use the gpb functions for inspecting what services are defined and build an ets table to map what the proto will send, `<package>.<Service>/<Method>` to an Erlang module and function.

Long story short.. hehe, yes, I want to still use snake casing from gpb in my messages but keep the service and rpcs as what grpc implementations will be sending.

**Comment by tomas-abrahamsson** [comment]
I see. Unfortunately, there are already some more options that have too wide meaning, for instance `msg_names_to_lower` (also implied by `epb_compatibility`) and I figure it won't be too popular to changing semantics of existing options.

So some alternatives: (a) a `preserve_names_for_gprpc` that effectively narrows the meaning of existing `msg_name_to_*` options, or (b) adding options with narrow names, like `just_msg_names_to_*`, `just_service_names_to_*` etc, and let `msg_name_to_*` expand to those with narrower names, or (c) something better, a name transformation function or something else? Naming is of course up for discussion. It is difficult coming up with names when the good ones are already taken. I'm not really sure what route to go at this point, maybe a bit inclined to (a), but with some better name, though option expansion is a bit easier to understand.

**Comment by tomas-abrahamsson** [comment]
Maybe a way to go could be to introduce `{name_control, [{msg_name, snake_case}, {enum_symbol, lowercase}, ...]}`.

Items to control could be `msg_name`, `field_name`, `enum_symbol`, `package_name`, `service_name`, `rpc_name` and more if we can think of anything more. Maybe also `module_name`.

Control could be `lowercase`, `snake_case`, `{prefix,Str}` or `{suffix,Str}`. Maybe also `preserve` (for an explicit as-is no-op). Maybe also a transformation function (but what args? -- maybe an expansion in some future)

For compatibility, the old options `msg_name_to_snake_case` could be seen as being expanded to or equivalent to `{name_control, [{msg_name, snake_case}, {service_name, snake_case}, {rpc_name, snake_case}]}`.

**Comment by tomas-abrahamsson** [comment]
Hi agian, I pushed a tentative renaming cleanup to the (a work-in-progress) branch, `renaming-options`. It introduces options `{rename,{What,How}}` where `What` can be `msg_fqname` for fully qualified message names, ie `Package.MsgName` or `What` can just `msg_name`, ie operating on only the message name, not the package name. Other `What`s are `pkg_name`, `service_fqname`, `service_name` and `rpc_name`.

The `How` can be `{prefix,Prefix}`, `{suffix,Suffix}`, `lowercase` or `snake_case`. A special form of `Prefix` can be `{by_proto,PrefixList}` just like previously.

On this branch, it is thus possible to say `{rename,{msg_name,snake_case}}` and it renames only the message name parts, it does not touch packages, services or rpcs.

The branch adds a module, `gpb_names`, which contains the renaming functionality (moved to here from the gpb_parse). This `gpb_names` module also exposes a functions for computing module names, and option renaming operations for those, so I think it should address your remaining concern in #121.  

The superfluous `_` after dot is also fixed.

This branch does not yet address anything in #124 for making nested message names prettier, but perhaps we can use this branch as a basis for what we want to do there.

For the new options, I'm of course open for suggestions, better names, reworkings, etc. Just to throw some ideas: an option for `How` to replacing dots with underscores (for #124), or an option for a user-supplied function for `How`, or maybe changing `{prefix,{by_proto,PrefixList}}` into `{prefix_by_proto,PrefixList}` to flatten things a bit.

**Comment by tomas-abrahamsson** [comment]
Sorry closed by mistake. Re-opened.

**Comment by tsloughter** [comment]
Great! I'll give this a shot this week.

**Comment by tsloughter** [comment]
I'm trying this out now. Definitely what I am looking for, thanks!

I'll close this issue, but I still have questions but they fit better in #124 :)
